File: GameEngine_IOS/GameEngine/Models/ComplexEntities/Material.swift
import Foundation

/**
 * Has all the required parameters to render one textured entity
 */
public class Material {
    
    /**
     * The identifier of the texture
     */
    var textureId : Int
    
    /**
     * The weight of the texture for the ambient color of the object
     */
    var textureWeight : Float
    
    /**
     * constant color of the ambient component
     */
    var diffuseColor : ColorRGBA
    
    /**
     * How damped the shine is
     */
    var shineDamper : Float
    
    /**
     * How reflective the model is
     */
    var reflectivity : Float
    
    /**
     * Indicates if the material has transparency or not
     */
    var hasTransparency : Bool;
    
    
    /**
     * Indicate that all the normals of the material are going to point up (in the
     * same direction
     */
    var normalsPointingUp : Bool;
    
    /**
     *Initiator of model texture
     *
     * @param textureId Identifier of the texture
     *
     */
    public init(_ textureId : Int) {
        self.textureId = textureId;
        self.shineDamper = 1.0;
        self.reflectivity = 0.0;
        self.hasTransparency = false
        self.normalsPointingUp = false
        self.textureWeight = 0.0
        self.diffuseColor = ColorRGBA(r: 1.0, g: 1.0, b: 1.0, a: 0.0)
    }
    
}

File: GameEngine_Android/app/src/main/java/com/dferreira/game_engine/models/complexEntities/Material.java
package com.dferreira.game_engine.models.complexEntities;

/**
 * Has the parameters of the material of a rawModel
 */
public class Material {

    /**
     * Diffuse component of the color
     */
    private LightingComponent diffuse;

    /**
     * How damped the shine is
     */
    private float shineDamper;

    /**
     * How reflective the model is
     */
    private float reflectivity;

    /**
     * Indicates if the model has transparency or not
     */
    private boolean hasTransparency;

    /**
     * Indicate that all our normals of the object are going to point up (in the
     * same direction
     */
    private boolean normalsPointingUp;

    /**
     * Constructor of the texture model
     */
    public Material() {
        this.shineDamper = 1.0f;
        this.reflectivity = 0.0f;
        this.hasTransparency = false;
        this.normalsPointingUp = false;
    }

    /**
     * @return the shineDamper
     */
    public float getShineDamper() {
        return shineDamper;
    }

    /**
     * @param shineDamper the shineDamper to set
     */
    @SuppressWarnings("SameParameterValue")
    public void setShineDamper(float shineDamper) {
        this.shineDamper = shineDamper;
    }

    /**
     * @return the reflectivity
     */
    public float getReflectivity() {
        return reflectivity;
    }

    /**
     * @param reflectivity the reflectivity to set
     */
    @SuppressWarnings("SameParameterValue")
    public void setReflectivity(float reflectivity) {
        this.reflectivity = reflectivity;
    }

    /**
     * @param hasTransparency If the model has any transparency or not
     */
    public void setHasTransparency(boolean hasTransparency) {
        this.hasTransparency = hasTransparency;
    }

    /**
     * Returns if the model has any transparency or not
     */
    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    public boolean hasTransparency() {
        return this.hasTransparency;
    }

    /**
     * @param normalsPointingUp If the normals of the model should point up
     */
    public void setNormalsPointingUp(boolean normalsPointingUp) {
        this.normalsPointingUp = normalsPointingUp;
    }

    /**
     * @return if the normals of the model should point up
     */
    public boolean areNormalsPointingUp() {
        return normalsPointingUp;
    }


    /**
     *
     * @return Get the diffuse component of the material
     */
    public LightingComponent getDiffuse() {
        return diffuse;
    }

    /**
     * Sets the diffuse component of the material
     *
     * @param diffuse The diffuse component to set
     */
    public void setDiffuse(LightingComponent diffuse) {
        this.diffuse = diffuse;
    }


}
